### PR TITLE
Add component abstractions

### PR DESCRIPTION
**Pull request by yufeih**
This is a draft abstraction of internal components used inside docfx. There are a couple of reasons to add this layer today:

- Clarify internal components, their boundary, responsibility and dependency.
- Ensures each functionality is implemented in one component to avoid bugs.
- Changes to one component does not affect other components
- Highlight cross component changes as they require changes to these interfaces and respective data contract.

This is a draft proposal, some of the abstractions match closely to what is already implemented in docfx, some other abstractions don't. We don't need to implement these interfaces immediately, it gives us a guidance to gradually morph into a better componentized codebase.

###### Microsoft Reviewers: [Open in CodeFlow](http://wpcp.azurewebsites.net/CodeFlowProtocolProxy2.php?pullrequest=https://github.com/dotnet/docfx/pull/5074)

### CODE REVIEW

**Comment by OsmondJiang** [comment]
That's what I really want at the beginning. How about we have a design review for that since it has a lot of impacts to our code future development. cc @herohua

File: src/docfx/abstractions/IRedirectionProvider.cs [hunk 1]
```diff
@@ -0,0 +1,12 @@
+// Copyright (c) Microsoft. All rights reserved.
+// Licensed under the MIT license. See LICENSE file in the project root for full license information.
+
+namespace Microsoft.Docs.Build
+{
+    internal interface IRedirectionProvider
```
Comment (mingwli) [inline src/docfx/abstractions/IRedirectionProvider.cs]: Maybe there's a lack of API to retrive sourcePath from renamedPath

File: src/docfx/abstractions/IBookmarkValidator.cs [hunk 2]
```diff
@@ -0,0 +1,12 @@
+// Copyright (c) Microsoft. All rights reserved.
+// Licensed under the MIT license. See LICENSE file in the project root for full license information.
+
+namespace Microsoft.Docs.Build
+{
+    internal interface IBookmarkValidator
+    {
+        void AddLink(FilePath declaringFile, FilePath targetFile, SourceInfo<string> url);
+
+        void AddBookmarks(FilePath file, string[] bookmarks);
```
Comment (OsmondJiang) [inline src/docfx/abstractions/IBookmarkValidator.cs]: Validate()?

File: src/docfx/abstractions/IRedirectionProvider.cs [hunk 3]
```diff
@@ -0,0 +1,12 @@
...
+    {
+        bool TryGetRedirection(string sourcePath, out string redirectUrl);
+
+        bool TryGetRename(string sourcePath, out string originalPath);
```
Comment (OsmondJiang) [inline src/docfx/abstractions/IRedirectionProvider.cs]: From the usage, TryGetRename is to get document id, should we directly provide TryGetRedirectionDocumentId()?

File: src/docfx/abstractions/IRepository.cs [hunk 4]
```diff
@@ -0,0 +1,16 @@
+// Copyright (c) Microsoft. All rights reserved.
+// Licensed under the MIT license. See LICENSE file in the project root for full license information.
+
+namespace Microsoft.Docs.Build
+{
+    internal interface IRepository
```
Comment (OsmondJiang) [inline src/docfx/abstractions/IRepository.cs]: IGitInfoProvider? It's conflicts with Repository data model

File: src/docfx/abstractions/IRestoreMap.cs [hunk 5]
```diff
@@ -0,0 +1,16 @@
+// Copyright (c) Microsoft. All rights reserved.
+// Licensed under the MIT license. See LICENSE file in the project root for full license information.
+
+namespace Microsoft.Docs.Build
+{
+    internal interface IRestoreMap
```
Comment (OsmondJiang) [inline src/docfx/abstractions/IRestoreMap.cs]: Can we keep File and Git splitted?  They are using totally different context,

File: src/docfx/abstractions/IRestoreMap.cs [hunk 6]
```diff
@@ -0,0 +1,16 @@
...
+    {
+        string ReadFile(SourceInfo<string> url);
+
+        string GetFilePath(SourceInfo<string> url);
+
+        (string path, string commit) GetGitRepositoryPath(string url, string branch);
```
Comment (OsmondJiang) [inline src/docfx/abstractions/IRestoreMap.cs]: PackageUrl

File: src/docfx/abstractions/ITableOfContentsResolver.cs [hunk 7]
```diff
@@ -0,0 +1,14 @@
+// Copyright (c) Microsoft. All rights reserved.
+// Licensed under the MIT license. See LICENSE file in the project root for full license information.
+
+namespace Microsoft.Docs.Build
+{
+    internal interface ITableOfContentsResolver
```
Comment (OsmondJiang) [inline src/docfx/abstractions/ITableOfContentsResolver.cs]: what's this for?

File: src/docfx/abstractions/ISchemaDocumentProvider.cs [hunk 8]
```diff
@@ -0,0 +1,14 @@
+// Copyright (c) Microsoft. All rights reserved.
+// Licensed under the MIT license. See LICENSE file in the project root for full license information.
+
+using Newtonsoft.Json.Linq;
+
+namespace Microsoft.Docs.Build
+{
+    internal interface ISchemaDocumentProvider
+    {
+        (Error[] errors, IXrefSpec[] xrefSpecs) LoadXrefSpecs(FilePath file);
```
Comment (OsmondJiang) [inline src/docfx/abstractions/ISchemaDocumentProvider.cs]: move it to IXrefResolver?

File: src/docfx/abstractions/ITemplateEngine.cs [hunk 9]
```diff
@@ -0,0 +1,22 @@
...
+
+namespace Microsoft.Docs.Build
+{
+    internal interface ITemplateEngine
+    {
+        bool IsPage(string schemaName);
+
+        bool TryGetLocalizedToken(string token, string localizedToken);
+
+        JsonSchema GetJsonSchema(SourceInfo<string> schemaName);
```
Comment (OsmondJiang) [inline src/docfx/abstractions/ITemplateEngine.cs]: Split it into IJsonSchemaProvider, shared it with IJsonSchemaValidator and IJsonSchemaTransformer?

**Comment by yufeih** [comment]
Abandon this as the purpose of this PR been served.

**Pull request closed without merging**